File: api/quakes.py
from __future__ import annotations

import json
import time
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler
from typing import Any
from urllib.error import URLError
from urllib.parse import parse_qs, urlparse
from urllib.request import Request, urlopen

BASE_URL = "https://earthquake.usgs.gov/earthquakes/feed/v1.0/summary/"
WINDOWS = {
    "hour": "all_hour",
    "day": "all_day",
    "week": "all_week",
}
CACHE_HEADERS = {
    "hour": "s-maxage=60, stale-while-revalidate=300",
    "day": "s-maxage=180, stale-while-revalidate=600",
    "week": "s-maxage=600, stale-while-revalidate=1800",
}


def _json_response(handler: BaseHTTPRequestHandler, payload: dict[str, Any], status: int) -> None:
    body = json.dumps(payload).encode("utf-8")
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json; charset=utf-8")
    handler.send_header("Content-Length", str(len(body)))
    handler.end_headers()
    handler.wfile.write(body)


def _parse_min_mag(value: str | None) -> float | None:
    if value is None:
        return None
    try:
        parsed = float(value)
    except ValueError:
        return None
    return parsed


def _fetch_feed(feed: str) -> dict[str, Any]:
    request = Request(f"{BASE_URL}{feed}.geojson", headers={"User-Agent": "EarthquakeConstellations/1.0"})
    with urlopen(request, timeout=10) as response:
        payload = response.read().decode("utf-8")
    return json.loads(payload)


def _normalize(feed: dict[str, Any], window: str, min_mag: float | None) -> dict[str, Any]:
    items = []
    for feature in feed.get("features", []):
        properties = feature.get("properties", {}) or {}
        geometry = feature.get("geometry", {}) or {}
        coords = geometry.get("coordinates", []) or []
        if len(coords) < 2:
            continue

        magnitude = properties.get("mag")
        if min_mag is not None and magnitude is not None and magnitude < min_mag:
            continue

        item = {
            "id": feature.get("id") or "",
            "time": properties.get("time") or 0,
            "mag": magnitude,
            "depthKm": coords[2] if len(coords) > 2 else None,
            "lon": coords[0],
            "lat": coords[1],
            "place": properties.get("place") or "Unknown location",
            "url": properties.get("url") or "",
        }
        items.append(item)

    bbox = feed.get("bbox")
    if not bbox or len(bbox) < 4:
        lons = [item["lon"] for item in items] or [0, 0]
        lats = [item["lat"] for item in items] or [0, 0]
        bbox = [min(lons), min(lats), max(lons), max(lats)]

    return {
        "generated": int(time.time() * 1000),
        "window": window,
        "count": len(items),
        "bbox": [bbox[0], bbox[1], bbox[2], bbox[3]],
        "items": items,
    }


class handler(BaseHTTPRequestHandler):
    def do_GET(self) -> None:
        parsed = urlparse(self.path)
        params = parse_qs(parsed.query)
        window = params.get("window", ["day"])[0]
        min_mag = _parse_min_mag(params.get("minMag", [None])[0])

        if window not in WINDOWS:
            _json_response(
                self,
                {"error": "Invalid window. Use hour, day, or week."},
                HTTPStatus.BAD_REQUEST,
            )
            return

        try:
            feed = _fetch_feed(WINDOWS[window])
        except (URLError, TimeoutError, json.JSONDecodeError):
            _json_response(
                self,
                {
                    "error": "USGS feed is temporarily unavailable. Please try again soon.",
                    "window": window,
                },
                HTTPStatus.SERVICE_UNAVAILABLE,
            )
            return

        payload = _normalize(feed, window, min_mag)
        body = json.dumps(payload).encode("utf-8")
        self.send_response(HTTPStatus.OK)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Cache-Control", CACHE_HEADERS[window])
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

File: lib/quakes.ts
import type { QuakeItem, QuakeResponse, QuakeWindow } from "@/types/quakes";

const WINDOW_SET = new Set<QuakeWindow>(["hour", "day", "week"]);

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === "object" && value !== null;

const isNumber = (value: unknown): value is number =>
  typeof value === "number" && Number.isFinite(value);

const isString = (value: unknown): value is string => typeof value === "string";

const isNullableNumber = (value: unknown): value is number | null =>
  value === null || isNumber(value);

export const windowMs: Record<QuakeWindow, number> = {
  hour: 60 * 60 * 1000,
  day: 24 * 60 * 60 * 1000,
  week: 7 * 24 * 60 * 60 * 1000
};

export const pollIntervalMs: Record<QuakeWindow, number> = {
  hour: 30_000,
  day: 120_000,
  week: 300_000
};

export function parseQuakeResponse(data: unknown): QuakeResponse {
  if (!isRecord(data)) {
    throw new Error("Invalid response payload");
  }

  const { generated, window, count, bbox, items } = data;

  if (!isNumber(generated)) {
    throw new Error("Invalid generated timestamp");
  }

  if (!isString(window) || !WINDOW_SET.has(window as QuakeWindow)) {
    throw new Error("Invalid window");
  }

  if (!isNumber(count)) {
    throw new Error("Invalid count");
  }

  if (!Array.isArray(bbox) || bbox.length !== 4 || !bbox.every(isNumber)) {
    throw new Error("Invalid bounding box");
  }

  if (!Array.isArray(items)) {
    throw new Error("Invalid items list");
  }

  const parsedItems: QuakeItem[] = items.map((item) => {
    if (!isRecord(item)) {
      throw new Error("Invalid item");
    }

    const { id, time, mag, depthKm, lon, lat, place, url } = item;

    if (!isString(id)) {
      throw new Error("Invalid id");
    }
    if (!isNumber(time)) {
      throw new Error("Invalid time");
    }
    if (!isNullableNumber(mag)) {
      throw new Error("Invalid magnitude");
    }
    if (!isNullableNumber(depthKm)) {
      throw new Error("Invalid depth");
    }
    if (!isNumber(lon) || !isNumber(lat)) {
      throw new Error("Invalid coordinates");
    }
    if (!isString(place)) {
      throw new Error("Invalid place");
    }
    if (!isString(url)) {
      throw new Error("Invalid url");
    }

    return { id, time, mag, depthKm, lon, lat, place, url };
  });

  return {
    generated,
    window: window as QuakeWindow,
    count,
    bbox: [bbox[0], bbox[1], bbox[2], bbox[3]],
    items: parsedItems
  };
}

export const clamp = (value: number, min: number, max: number) =>
  Math.min(max, Math.max(min, value));
